File: server/internal/store/puzzle.go
package store

import (
	"context"
	"encoding/json"

	"github.com/jackc/pgx/v5"
)

func (s *PgStore) CreatePuzzle(ctx context.Context, p *Puzzle) (*Puzzle, error) {
	b, _ := json.Marshal(p.WordSet)
	row := s.pool.QueryRow(ctx, `
		INSERT INTO puzzles (owner_id, category_ids, difficulty, timer_seconds, word_set)
		VALUES ($1, $2, $3, $4, $5)
		RETURNING id, created_at
	`, p.OwnerID, p.CategoryIDs, p.Difficulty, p.TimerSeconds, b)
	if err := row.Scan(&p.ID, &p.CreatedAt); err != nil {
		return nil, err
	}
	return p, nil
}

func (s *PgStore) GetPuzzleByID(ctx context.Context, id string) (*Puzzle, error) {
	row := s.pool.QueryRow(ctx, `
		SELECT id, owner_id, category_ids, difficulty, timer_seconds, word_set, created_at
		FROM puzzles WHERE id = $1
	`, id)
	return scanPuzzle(row)
}

func (s *PgStore) GetPuzzlesByOwner(ctx context.Context, ownerID string) ([]Puzzle, error) {
	rows, err := s.pool.Query(ctx, `
		SELECT id, owner_id, category_ids, difficulty, timer_seconds, word_set, created_at
		FROM puzzles WHERE owner_id = $1 ORDER BY created_at DESC
	`, ownerID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []Puzzle
	for rows.Next() {
		p, err := scanPuzzle(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, *p)
	}
	return out, rows.Err()
}

func scanPuzzle(row pgx.Row) (*Puzzle, error) {
	var p Puzzle
	var raw []byte
	err := row.Scan(&p.ID, &p.OwnerID, &p.CategoryIDs, &p.Difficulty, &p.TimerSeconds, &raw, &p.CreatedAt)
	if err != nil {
		return nil, err
	}
	if raw != nil {
		_ = json.Unmarshal(raw, &p.WordSet)
	}
	return &p, nil
}

File: mobile/lib/models/puzzle.dart
class PuzzleWord {
  final String word;
  final String correctCategory;
  final List<String> decoys;

  PuzzleWord({
    required this.word,
    required this.correctCategory,
    required this.decoys,
  });

  Map<String, dynamic> toJson() => {
    'word': word,
    'correct_category': correctCategory,
    'decoys': decoys,
  };

  factory PuzzleWord.fromJson(Map<String, dynamic> json) => PuzzleWord(
    word: json['word'] as String,
    correctCategory: json['correct_category'] as String,
    decoys: List<String>.from(json['decoys'] as List),
  );
}

class Puzzle {
  final String? id;
  final String? ownerId;
  final List<String> categoryIds;
  final int difficulty;
  final int timerSeconds;
  final List<PuzzleWord> wordSet;

  Puzzle({
    this.id,
    this.ownerId,
    required this.categoryIds,
    required this.difficulty,
    required this.timerSeconds,
    required this.wordSet,
  });

  Map<String, dynamic> toJson() => {
    'category_ids': categoryIds,
    'difficulty': difficulty,
    'timer_seconds': timerSeconds,
    'word_set': wordSet.map((w) => w.toJson()).toList(),
  };

  factory Puzzle.fromJson(Map<String, dynamic> json) => Puzzle(
    id: json['id'] as String?,
    ownerId: json['owner_id'] as String?,
    categoryIds: List<String>.from(json['category_ids'] as List? ?? []),
    difficulty: json['difficulty'] as int? ?? 1,
    timerSeconds: json['timer_seconds'] as int? ?? 50,
    wordSet: (json['word_set'] as List? ?? [])
      .map((e) => PuzzleWord.fromJson(e as Map<String, dynamic>))
      .toList(),
  );
}
